### PR TITLE
docs: Add guide for testing dataset definitions

### LINKED ISSUE #1830: Write how-to guide for assure command
This should be brief.  If something is hard to explain, it might mean that the feature could be improved!

### PR DESCRIPTION
**Pull request by milanwiedemann**
This documents the current process and will be continuously improved together with improvements of the `assure` command.

Closes #1830

### CODE REVIEW

File: docs/how-to/index.md [hunk 1]
```diff
@@ -6,3 +6,4 @@ The how-to guides provide practical steps for working with ehrQL in your project
 * [How to use dummy data in an ehrQL measures definition](dummy-measures-data.md)
 * [How to assign multiple columns to a dataset programmatically](assign-multiple-columns.md)
 * [How to work with codelists](codelists.md)
+* [How to test the ehrql logic in your dataset definition](test-dataset-definition.md)
```
Comment (milanwiedemann) [inline docs/how-to/index.md:9]: Oh, this was the original _longer_ title, will fix this together with suggestions that come up during code review

File: docs/how-to/test-dataset-definition.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    .last_for_patient()
+)
+
+dataset.has_asthma_med = latest_asthma_med.exists_for_patient()
+dataset.latest_asthma_med_date = latest_asthma_med.date
+```
+
+## Specifying data and expectations
+
+Next, you need to provide (1) data for test patients and (2) define the expected outcomes for these patients after applying your ehrQL queries.
```
Comment (inglesp) [inline docs/how-to/test-dataset-definition.md:47]: "expected outcomes" doesn't quite capture it.

How about

> (2) specify the data that you expect to see in the dataset for each patient

?

File: docs/how-to/test-dataset-definition.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+Test data and expectations are defined in a nested dictionary called `patient_data`.
+The key in the outermost dictionary specifies the patient id.
+
+### Data for test patients
+
+In the example below we have created one test patient with the patient id `1` and specified more data for this patient in two lists called `patient` and `medications`.
+The names of the inner lists match the names of the tables you are using in your dataset definition.
+To explore how these tables are structured you can look at the column names in the [core schema](../../reference/schemas/core) documentation.
+You can also generate CSV files of these tables and add them to your project using the [dump-example-data](../../reference/cli/#dump-example-data) or [create-dummy-tables](../../reference/cli/#create-dummy-tables) commands.
```
Comment (inglesp) [inline docs/how-to/test-dataset-definition.md:56]: Does talking about dummy data complicate things here?

File: docs/how-to/test-dataset-definition.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+The `patients` list contains one dictionary with two keys, one key for each column (`date_of_birth` and `sex`) that we want to create.
```
Comment (inglesp) [inline docs/how-to/test-dataset-definition.md:58]: How about "populate" instead of "create"?

File: docs/how-to/test-dataset-definition.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+Note that you don't have to specify a value for each column in the underlying table.
+For example we did not specify `date_of_death` in the dictionary inside the `patients` list so the column will be missing with the value `None`.
+
+The `medications` list contains two dictionaries (one for each row we add to the table) with two keys (one for each column: `date` and `dmd_code`).
+Note that you have to specify a list for each table you import from `ehrql.tables.core` in your dataset definition, but this could also be an empty list.
```
Comment (inglesp) [inline docs/how-to/test-dataset-definition.md:63]: Not quite true: you have to specify a list for each table you use in your dataset definition.  You could import a table and then not use it.

File: docs/how-to/test-dataset-definition.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+* **Patient 1**: Expected in our population because the patient is older than 18 years on the `index_date`.
+  The three entries in the medications table tests the ehrQL logic that selects the latest medication before the `index_date`.
+* **Patient 2**: Expected in our population because the patient is older than 18 years on the `index_date`.
+  However the patient does not have any entries in their `medications` table.
+  Here we are testing the behaviour of our ehrQL query when a patient was never prescribed a code from the `asthma_codes` codelist
+* **Patient 3**: Not expected in our population because the patient is younger than 18 years on the `index_date`.
+
+```py
+from datetime import date
+from dataset_definition import dataset
```
Comment (inglesp) [inline docs/how-to/test-dataset-definition.md:108]: You need to mention why this dataset has to be imported in the test file.

**Review by inglesp — APPROVED** [review]
Lovely clear technical writing, well done.

Have made a handful of small suggestions.

File: docs/how-to/test-dataset-definition.md (hunk 3 above)
Comment (milanwiedemann) [reply docs/how-to/test-dataset-definition.md:56]: probably yes, I think refering to the schema is helpful enough!

File: docs/how-to/test-dataset-definition.md (hunk 5 above)
Comment (milanwiedemann) [reply docs/how-to/test-dataset-definition.md:63]: thanks, that makes even simpler to describe!

**Pull request merged**